src/Engine/ProtoCore/AssociativeGraph.cs

```diff
@@ -1527,8 +1527,20 @@ public DependencyGraph(Core core)
             graphNodeMap = new Dictionary<ulong, List<GraphNode>>();
         }
 
+        private List<GraphNode> cachedNodes;
+
         public List<GraphNode> GetGraphNodesAtScope(int classIndex, int procIndex)
         {
+            if (classIndex == Constants.kInvalidPC && procIndex == Constants.kInvalidPC)
+            {
+                if (cachedNodes == null)
+                {
+                    graphNodeMap.TryGetValue(GetGraphNodeKey(classIndex, procIndex), out cachedNodes);
+                }
+
+                return cachedNodes;
+            }
```

Comment:
Dictionary lookup is basically constant time, so how much time does caching really save?

Reply:
I think it is about scale. This is a nested dictionary lookup as there is another dictionary look up in `GetGraphNodeKey`.  In my sample graph `GetGraphNodesAtScope` is called 3 million times which adds up.